// htmlbars/dist/cjs/htmlbars-runtime.js
exports.__esModule = true;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var _htmlbarsRuntimeHooks = require('./htmlbars-runtime/hooks');

var _htmlbarsRuntimeHooks2 = _interopRequireDefault(_htmlbarsRuntimeHooks);

var _htmlbarsRuntimeRender = require('./htmlbars-runtime/render');

var _htmlbarsRuntimeRender2 = _interopRequireDefault(_htmlbarsRuntimeRender);

var _htmlbarsUtilMorphUtils = require("./htmlbars-util/morph-utils");

var _htmlbarsUtilTemplateUtils = require("./htmlbars-util/template-utils");

var internal = {
  blockFor: _htmlbarsUtilTemplateUtils.blockFor,
  manualElement: _htmlbarsRuntimeRender.manualElement,
  hostBlock: _htmlbarsRuntimeHooks.hostBlock,
  continueBlock: _htmlbarsRuntimeHooks.continueBlock,
  hostYieldWithShadowTemplate: _htmlbarsRuntimeHooks.hostYieldWithShadowTemplate,
  visitChildren: _htmlbarsUtilMorphUtils.visitChildren,
  validateChildMorphs: _htmlbarsUtilMorphUtils.validateChildMorphs,
  clearMorph: _htmlbarsUtilTemplateUtils.clearMorph
};

exports.hooks = _htmlbarsRuntimeHooks2.default;
exports.render = _htmlbarsRuntimeRender2.default;
exports.internal = internal;
